File: IPWorks Samples/CalDAV Client/python/caldav_client.py
# 
# IPWorks 2022 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of IPWorks in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/ipworks
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from ipworks import *

input = sys.hexversion<0x03000000 and raw_input or input

def ensureArg(args, prompt, index):
  if len(args) <= index:
    while len(args) <= index:
      args.append(None)
    args[index] = input(prompt)
  elif args[index] == None:
    args[index] = input(prompt)

caldav = CalDAV()
oauth = OAuth()

resource_uris = []

def fire_SSL_server_authentication(e):
    e.accept = True
	
def event_details(e):
    resource_uris.append(e.resource_uri)
    s = "%2d) %-50.50s %-20s%-30.30s %-100s\n" % (len(resource_uris), caldav.get_summary(), caldav.get_start_date(), caldav.get_location(),e.resource_uri)
    print(s)

if len(sys.argv) < 3:
    print("usage: caldav [options] provider username password")
    print("Options")
    print("  -i        OAuth Client ID")
    print("  -s        OAuth Client Secret")
    print("  provider  the provider name: Google ")
    print("  username  the username to login")
    print("  password  the password to login")
    print("\r\nExample: python cal_client.py -i \"Client ID\" -s \"Client Secret\" google username password")
else:
    try:
        

        for i in range(0,len(sys.argv)):
            if (sys.argv[i].startswith("-")):
                if (sys.argv[i]=="-i"):
                    oauth.set_client_id(sys.argv[i+1])
                if (sys.argv[i]=="-s"):
                    oauth.set_client_secret(sys.argv[i+1])

        
        caldav.on_ssl_server_authentication = fire_SSL_server_authentication
        # caldav.on_status = fire_Status
        caldav.on_event_details = event_details
        oauth.set_server_auth_url("https://accounts.google.com/o/oauth2/auth")
        oauth.set_server_token_url("https://oauth2.googleapis.com/token")
        caldav.set_auth_scheme(6)
        
        while True:
            command = input("Choose an option:\n1) List Events\n2) Add Event\n3) Delete Event\nQ) Quit.\n> ")
            if command == "1":
                caldav.reset()
                resource_uris[:] = [] # Clear the list
                event_num = 1
                caldav.set_user(sys.argv[len(sys.argv) - 2])
                caldav.set_password(sys.argv[len(sys.argv) - 2])
                oauth.set_authorization_scope("https://www.googleapis.com/auth/calendar")
                auth=oauth.get_authorization()
                caldav.set_authorization(auth)
                print("     %-50.50s %-20s%-30.30s %-100s\n" % ("Summary","Start Date","Location","ResourceURI"))
                s = "https://apidata.googleusercontent.com/caldav/v2/%s/events" % (caldav.get_user())
                caldav.get_calendar_report(s)

            elif command == "2":
                caldav.reset()
                # start = 
                # end = str(input("End Date (YYYYMMDDTHHMMSS): "))
                # location = str(input("Location: "))
                # summary = str(input("Summary: "))
                # desc = str(input("Description: "))
                caldav.set_user(sys.argv[len(sys.argv) - 2])
                caldav.set_password(sys.argv[len(sys.argv) - 2])
                oauth.set_authorization_scope("https://www.googleapis.com/auth/calendar")

                caldav.set_start_date(str(input("Start Date (YYYYMMDDTHHMMSS): ")))
                caldav.set_end_date(str(input("End Date (YYYYMMDDTHHMMSS): ")))
                caldav.set_location(str(input("Location: ")))
                caldav.set_summary(str(input("Summary: ")))
                caldav.set_description(str(input("Description: ")))
                caldav.set_uid(caldav.get_start_date())
                caldav.set_authorization(oauth.get_authorization())
                caldav.put_calendar_event("https://apidata.googleusercontent.com/caldav/v2/" + caldav.get_user() + "/events/" + caldav.get_uid() + ".ics")
                print("Event successfully added.")

            elif command == "3":
                caldav.reset()
                caldav.set_user(sys.argv[len(sys.argv) - 2])
                caldav.set_password(sys.argv[len(sys.argv) - 2])
                oauth.set_authorization_scope("https://www.googleapis.com/auth/calendar")
                index = int(input("Specify the number of the event to delete: "))
                if index > len(resource_uris):
                    print("\nPlease run 1) List Events command first.\n")
                    continue
                caldav.set_authorization(oauth.get_authorization())
                caldav.delete_calendar_event("https://apidata.googleusercontent.com" + resource_uris[index - 1])
                print("Event successfully deleted.")
               


           
            elif command == "q" or command == "Q":
                sys.exit()
        
    except IPWorksError as e:
        print("ERROR %s" %e)

File: IPWorks Samples/CalDAV Client/php/caldav_client.php
<?php
/*
 * IPWorks 2024 PHP Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
require_once('../include/ipworks_caldav.php');
require_once('../include/ipworks_oauth.php');
require_once('../include/ipworks_const.php');
?>
<?php
require_once "../include/ipworks_caldav.php";
require_once "../include/ipworks_oauth.php";

class MyDAV extends IPWorks_CalDAV
{
  public function fireSSLServerAuthentication($param){
      $param["accept"] = true;
      return $param;
  }
  public function fireEventDetails($param){
    $startDateFormatted = date("m/d H:i:s", strtotime($this->getStartDate()));
    echo sprintf("%-40s %-20s %-30s %-100s\n",
      $this->getSummary(),
      $startDateFormatted,
      $this->getLocation(),
      $param["resourceuri"]
    );
  }
}

function prompt($message) {
  echo $message . ": ";
  return trim(fgets(STDIN));
}

function authorize(){
  global $caldav, $oauth, $authString, $user;
  $caldav->setUser($user);
  if(empty($authString)){
    $authString = $oauth->doGetAuthorization();
  }
  $caldav->setAuthorization($authString);
}

if (count($argv) < 4) {
  echo "usage: php caldav.php [options] provider username password\n";
  echo "Options:\n";
  echo "  -i        OAuth Client ID\n";
  echo "  -s        OAuth Client Secret\n";
  echo "  provider  the provider name: Google\n";
  echo "  username  the username to login\n";
  echo "  password  the password to login\n";
  echo "\nExample: php caldav.php -i \"clientID\" -s \"clientSecret\" google username password\n";
  exit(1);
}

echo "******************************************************************************\n";
echo "* This demo shows how to use the CalDAV component to list upcoming events    *\n";
echo "* from an existing Google calendar. You can also create a new event,         *\n";
echo "* delete an event, or export an event to an ics file.                        *\n";
echo "******************************************************************************\n\n";

$caldav = new MyDAV();
$oauth = new IPWorks_OAuth();

$clientID = "";
$clientSecret = "";
$user = $argv[count($argv) - 2];
$authString = "";

for ($i = 0; $i < $argc; $i++) {
  if ($argv[$i][0] === '-') {
    if ($argv[$i] === '-i') $clientID = $argv[$i + 1];
    if ($argv[$i] === '-s') $clientSecret = $argv[$i + 1];
  }
}
$oauth->setClientId($clientID);
$oauth->setClientSecret($clientSecret);
$oauth->setServerAuthURL("https://accounts.google.com/o/oauth2/auth");
$oauth->setServerTokenURL("https://accounts.google.com/o/oauth2/token");
$oauth->setAuthorizationScope("https://www.googleapis.com/auth/calendar");

echo "Type \"?\" for a list of commands.\n";

while (true) {
  echo "> ";
  $command = trim(fgets(STDIN));
  $args = explode(' ', $command);

  switch ($args[0]) {
    case '?':
      echo "1) List Events\n";
      echo "2) Add Event\n";
      echo "3) Delete Event\n";
      echo "4) Export ICS File\n";
      echo "?) This help menu.\n";
      echo "Q) Quit.\n";
      break;

    case '1': // List Calendar Events
      authorize();
      echo sprintf("%-40s %-20s %-30s %-100s\n",
          "Summary",
          "Start date",
          "Location",
          "ResourceUri"
      );
      $caldav->setReportFilterEventType(0);
      $caldav->doGetCalendarReport("https://apidata.googleusercontent.com/caldav/v2/" . $user . "/events");
      break;
      
    case '2': // Add Event
      $caldav->doReset();
      authorize();

      $startDate = prompt("Start Date (MM/dd/yyyy HH:mm:ss)");
      $endDate = prompt("End Date (MM/dd/yyyy HH:mm:ss)");
      $location = prompt("Location");
      $summary = prompt("Summary");
      $description = prompt("Description");
      $caldav->setStartDate(date("Ymd\THis", strtotime($startDate)));
      $caldav->setEndDate(date("Ymd\THis", strtotime($endDate)));
      $caldav->setLocation($location);
      $caldav->setSummary($summary);
      $caldav->setDescription($description);
      $caldav->setUID($caldav->getStartDate());
      $caldav->doCreateEvent("https://apidata.googleusercontent.com/caldav/v2/" . $user . "/events/" . $caldav->getUID() . ".ics");

      echo "Event Successfully Added.\n";
      break;

    case "3": //Delete event
      $caldav->doReset();
      $file = prompt("Give name of file to delete (find the file names by running option #1)");
      authorize();
      $caldav->doDeleteEvent("https://apidata.googleusercontent.com/caldav/v2/" . $user . "/events/" . $file);
      echo "Event Successfully Deleted.\n";
      break;

    case "4": //Export ICS
      $caldav->doReset();
      $file = prompt("Give name of .ICS file");
      authorize();
      $caldav->doGetEvent("https://apidata.googleusercontent.com/caldav/v2/" . $user . "/events/" . $file);
      $icsContent = $caldav->doExportICS();
      file_put_contents($file . '.ics', $icsContent);
      echo "ICS file saved.\n";
      break;

    case 'q':
    case 'Q':
      exit(0);

    default:
      echo "Bad command / Not implemented in demo.\n";
      break;
  }
}